### PR TITLE
Return logical vectors from predicate methods

### LINKED ISSUE #251: Generic predicates should operate on proxy but not restore
They probably should always return an untyped bool.

```r
new_cached_sum <- function(x = double(), sum = 0L) {
  new_vctr(x, sum = sum, class = "vctrs_cached_sum")
}
vec_restore.vctrs_cached_sum <- function(x, to, ..., i = NULL) {
  new_cached_sum(x, sum(x))
}
x <- new_cached_sum(runif(10))

is.finite(x)
#> <vctrs_cached_sum[10]>
#>  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
```

### PR DESCRIPTION
**Pull request by lionel-**
Closes #251

This still goes through `vec_math()` because a `vec_logical()` might not make sense for `is.finite()` or `is.nan()` which are defined on numeric data. Since we need to special-case them in `vec_math()`, a new generic just for `any()` and `all()` seems a bit overkill.

### CODE REVIEW

File: R/numeric.R [hunk 1]
```diff
@@ -42,15 +42,17 @@ vec_math <- function(fun, x, ...) {
...
-  } else {
-    # nocov start
+  if (!is_double(x) && !(is_logical(x) && fun %in% c("any", "all"))) {
     stop_unimplemented(x, "vec_math")
-    # nocov end
   }
+
+  out <- vec_math_base(fun, x, ...)
+
+  if (!fun %in% c("any", "all", "is.finite", "is.infinite", "is.nan")) {
```
Comment (hadley) [inline R/numeric.R]: What if we just restore if `out` is also double?

File: R/numeric.R [hunk 2]
```diff
@@ -42,15 +42,17 @@ vec_math <- function(fun, x, ...) {
 
 #' @export
 vec_math.default <- function(fun, x, ...) {
-  if (is_double(x)) {
-    vec_restore(vec_math_base(fun, x, ...), x)
-  } else if (fun %in% c("any", "all") && is_logical(x)) {
-    vec_restore(vec_math_base(fun, x, ...), x)
-  } else {
-    # nocov start
+  if (!is_double(x) && !(is_logical(x) && fun %in% c("any", "all"))) {
```
Comment (hadley) [inline R/numeric.R]: If found this compound expression rather hard to parse.

File: tests/testthat/test-type-vctr.R [hunk 3]
```diff
@@ -438,8 +438,8 @@ test_that("na.rm is forwarded to summary generics", {
 
   x <- new_vctr(lgl(TRUE, NA))
 
-  expect_identical(all(x, na.rm = FALSE), new_vctr(lgl(NA)))
-  expect_identical(all(x, na.rm = TRUE), new_vctr(TRUE))
+  expect_identical(all(x, na.rm = FALSE), lgl(NA))
```
Comment (hadley) [inline tests/testthat/test-type-vctr.R]: This is a change, so it's worth giving it a news bullet.

File: R/numeric.R (hunk 1 above)
Comment (lionel-) [reply R/numeric.R]: This also works, done.
